caas/kubernetes/provider/exec/copy.go

```diff
@@ -0,0 +1,220 @@
+// Copyright 2019 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package exec
+
+import (
+	"archive/tar"
+	"bytes"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+
+	"github.com/juju/errors"
+)
+
+// FileResource holds all the necessary parameters for the resource or destination of copy request.
+type FileResource struct {
+	Path          string
+	PodName       string
+	ContainerName string
+}
+
+func (cp *FileResource) validate() (err error) {
+	if cp.Path == "" {
+		return errors.New("path was missing")
+	}
+	return nil
+}
+
+// CopyParam holds all the necessary parameters for a copy request.
+type CopyParam struct {
+	Src                            FileResource
+	Dest                           FileResource
+	noPreserveOwnershipPermissions bool
+}
+
+func (cp *CopyParam) validate() error {
+	if err := cp.Src.validate(); err != nil {
+		return errors.Trace(err)
+	}
+	if err := cp.Dest.validate(); err != nil {
+		return errors.Trace(err)
+	}
+	if cp.Src.PodName != "" && cp.Dest.PodName != "" {
+		return errors.New("cross pods copy is not supported")
+	}
+	if cp.Src.PodName == "" && cp.Dest.PodName == "" {
+		return errors.New("copy either from pod nor to pod")
+	}
+	return nil
+}
+
+// Exec copy files/directories from host to a pod or from a pod to host.
+func (c client) Copy(params CopyParam, cancel <-chan struct{}) error {
+	if err := params.validate(); err != nil {
+		return errors.Trace(err)
+	}
+	if params.Src.PodName != "" {
+		return c.copyFromPod(params, cancel)
+	}
+	if params.Dest.PodName != "" {
+		return c.copyToPod(params, cancel)
+	}
+	return nil
+}
+
+func (c client) copyFromPod(params CopyParam, cancel <-chan struct{}) error {
+	// TODO(caas): implement when we need later.
+	return errors.NotSupportedf("copy from pod")
+}
+
+// this is inspired by kubectl cmd package.
+// - https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/cp/cp.go
+func (c client) copyToPod(params CopyParam, cancel <-chan struct{}) (err error) {
+	src := params.Src
+	dest := params.Dest
+	logger.Debugf("coping from %v to %v", src, dest)
+
+	if _, err = os.Stat(src.Path); err != nil {
+		return errors.NewNotValid(nil, fmt.Sprintf("%q does not exist on local", src.Path))
+	}
+
+	if dest.Path != "/" && strings.HasSuffix(dest.Path, "/") {
+		dest.Path = strings.TrimSuffix(dest.Path, "/")
+	}
+
+	if err = c.checkRemotePathIsDir(dest, cancel); err == nil {
+		dest.Path = path.Join(dest.Path, path.Base(src.Path))
+	}
+
+	reader, writer := c.pipGetter()
+
+	go func() {
+		defer writer.Close()
+		err = makeTar(src.Path, dest.Path, writer)
+		if err != nil {
+			logger.Errorf("make tar %q failed: %v", src.Path, err)
+		}
+	}()
+
+	cmds := []string{"tar", "-xmf", "-"}
```

Comment:
We'll need to support copy without tar. Most containers should have GNU tar or busybox, but from scratch might not. Possibly the solution to this is to precopy busybox-static to the target pod and use it for shell, tar etc. Initial tools will need to be static ELFs, as we can't rely on libc or even which c lib is installed.

Reply:
yeah, a good point.
But we might not have the ability to run any `.py` or even `jujud` inside `scratch`.
The image can run actions that should have some specific dependencies installed already.
But we definitely should consider enhancing here to make it more generic later.

Reply:
Added a card for this https://trello.com/c/CZLkihZ3